server/jetstream_api.go

```diff
@@ -1012,30 +1012,121 @@ func (s *Server) sendAPIErrResponse(ci *ClientInfo, acc *Account, subject, reply
 
 const errRespDelay = 500 * time.Millisecond
 
-func (s *Server) sendDelayedAPIErrResponse(ci *ClientInfo, acc *Account, subject, reply, request, response string, rg *raftGroup) {
-	js := s.getJetStream()
-	if js == nil {
+type delayedAPIResponse struct {
+	ci       *ClientInfo
+	acc      *Account
+	subject  string
+	reply    string
+	request  string
+	response string
+	rg       *raftGroup
+	deadline time.Time
+	next     *delayedAPIResponse
+}
+
+// Add `r` in the list that is maintained ordered by the `delayedAPIResponse.deadline` time.
+func addDelayedResponse(head, tail **delayedAPIResponse, r *delayedAPIResponse) {
+	// Check if list empty.
+	if *head == nil {
+		*head, *tail = r, r
 		return
 	}
-	var quitCh <-chan struct{}
-	js.mu.RLock()
-	if rg != nil && rg.node != nil {
-		quitCh = rg.node.QuitC()
+	// Check if it should be added at the end, which is if after or equal to the tail.
+	if r.deadline.After((*tail).deadline) || r.deadline.Equal((*tail).deadline) {
+		(*tail).next, *tail = r, r
+		return
 	}
-	js.mu.RUnlock()
+	// Find its spot in the list.
+	var prev *delayedAPIResponse
+	for c := *head; c != nil; c = c.next {
+		// We insert only if we are stricly before the current `c`.
+		if r.deadline.Before(c.deadline) {
+			r.next = c
+			if prev != nil {
+				prev.next = r
+			} else {
+				*head = r
+			}
+			return
+		}
+		prev = c
+	}
+}
 
-	s.startGoRoutine(func() {
-		defer s.grWG.Done()
+func (s *Server) delayedAPIResponder() {
+	defer s.grWG.Done()
+	var (
+		rgQuitCh   <-chan struct{}
+		head, tail *delayedAPIResponse
+		r          *delayedAPIResponse
+		tm         = time.NewTimer(time.Hour)
+	)
+	setup := func() {
+		r, rgQuitCh = nil, nil
+		// Check that JetStream is still on. Do not exit the go routine
+		// since JS can be enabled/disabled. The go routine will exit
+		// only if server is shutdown.
+		js := s.getJetStream()
```

Comment:
The check for JS and the raft node quit channel, although working and the test checks that, may be a bit of an overkill. Does it really matter if a delayed response is sent or not after a raft group is shutdown or JS turned off? Again, code is there and test verifies it works, but we may have done without I think.

Reply:
I think it makes sense that we don't send anything back after the Raft group is stopped, otherwise it might be a signal for the client/application to "try again" rather than just timing out in the rate-limited scenario.